### ISSUE TITLE
feature: improve and unify pagination and filtered pagination

### ISSUE THREAD

**Issue by elboletaire** [vocdoni/vocdoni-node#1320]
**Describe the feature**
Improve and unify all endpoints with pagination.

**Motivation**
Current pagination implementation has some issues and limitations.

**Proposal**
To have, preferably via query params, a way to filter all paginated results by indexed fields. 

Current filtering implementation has a specific POST endpoint for filtering results... that's one of the things that could be improved by just using GET params on its original endpoint.

For example, [list elections filtered](https://developer.vocdoni.io/vocdoni-api/list-elections-filtered) should not exist and, instead, [list elections](https://developer.vocdoni.io/vocdoni-api/list-elections) should implement the filtering via query params.

Taking that as an example:
- `/elections` should be the only URL to get the elections filtered and paginated
- Any indexed field should be a possible GET param to filter, i.e. `?finalResults=true&manuallyEnded=true`
- `limit` and `page` may be reserved words for pagination purposes. `order` too, but that's a bit more complicated since you need to specify the direction (asc/desc).
- the results should return, among the current `processes` key, other keys with information like the total results with the applied filter

Example request with pagination and filtering:

~~~
https://api-dev.vocdoni.net/v3/elections?page=1&limit=20&finalResults=true
~~~

Expected response:

~~~js
{
  "elections": [
    // a maximum of 20 elections with finalResults
    {
      // election body
    },
   "total": 140 // the total number of filtered results
}
~~~

Other information like the page or the limit could also be added to this response but, considering is info given by the requester... I don't see it as a requirement. The important one here is `total`.

**Affected routes**

Important ones:
- [List elections](https://developer.vocdoni.io/vocdoni-api/list-elections) (and [filtered](https://developer.vocdoni.io/vocdoni-api/list-elections-filtered))
- [List organizations](https://developer.vocdoni.io/vocdoni-api/list-organizations) (and [filtered](https://developer.vocdoni.io/vocdoni-api/list-organizations-filtered))
- [List organization elections](https://developer.vocdoni.io/vocdoni-api/list-organization-elections) (to be unified with [list organizations by status](https://developer.vocdoni.io/vocdoni-api/list-organization-elections-by-status) maybe)

Not so important, but should also be unified for coherence:
- [Transactions in a block](https://developer.vocdoni.io/vocdoni-api/transactions-in-a-block)
- [List transactions](https://developer.vocdoni.io/vocdoni-api/list-transactions) (maybe to be merged with [by index](https://developer.vocdoni.io/vocdoni-api/transaction-by-index) and [by hash](https://developer.vocdoni.io/vocdoni-api/transaction-by-hash))
- [List of existing accounts](https://developer.vocdoni.io/vocdoni-api/list-of-the-existing-accounts)
- [List account token fees](https://developer.vocdoni.io/vocdoni-api/list-account-token-fees)
- [List account received and sent token transfers](https://developer.vocdoni.io/vocdoni-api/list-account-received-and-sent-token-transfers)
- [List all token fees](https://developer.vocdoni.io/vocdoni-api/list-all-token-fees) (and also maybe to be unified with [list token fees by reference](https://developer.vocdoni.io/vocdoni-api/list-all-token-fees-by-reference), and [by type](https://developer.vocdoni.io/vocdoni-api/list-all-token-fees-by-type))
- [List election votes](https://developer.vocdoni.io/vocdoni-api/list-election-votes)

> Note: I've added every single call I've seen requiring a `:page` param (which now should be GET params). I've also set some as "maybe to be merged with" because seems like we have different calls for filtering for different values, and all these could be unified in the same call by simply specifying the indexed field in the URL as a GET param.

Task #1318 is related since after having that, those new stored metadata fields could/should be filterable via pagination.

**Comment by p4u** [comment]
It is important to note that @elboletaire is proposing a /v3/ API. So the /v2/ must be kept for backwards compatibility. 

In this v3 version, we are using query parameters, which were not part of the v2. So instead of `/v2/whatever/page/2`
we would do `/v3/whatever?page=2`

Tbh this is a big change to the API design, we should evaluate if it is worth it (cost of change VS what we are winning).

Another approach would be to keep using the same basis as v2, keeping always backwards compatibility, but:

1. Allow search parameters on JSON body (this is currently the way we do it)
2. Standardize the pagination on the backend side using /page/:number and providing "remainingPages" on the response

```json
GET /v2/elections/page/1
{
  "finalResults": true,
  "titleSearchTerm": "vocdoni"
}
``` 

```json
RESPONSE
{
 "elections": [...],
 "remainingPages": 9
}
```

So, I think the question is either:
- A: we create /v3/ with a design based on query parameters
- B: we keep /v2/ but standardize pagination and add new filter body fields

What do you think @mvdan @altergui @marcvelmer @elboletaire

**Comment by elboletaire** [comment]
Yes I think that may be a better (and not so breaking) approach.

But still, I'd suggest to try to add some of the functionalities to some of the basic calls though, like with `GET /elections`, we could still add the full specification I propose to that method, maintaining its current functionality (if no query params are received, fallbacks to its current functionality).

So I'd add the functionality as you just described @p4u:

`GET /v2/elections/page/1?finalResults=true&titleSearchTerm=vocdoni&limit=20`
~~~js
{
 "elections": [...],
 "remainingPages": 9
}
~~~


But would also add the new functionality to:

`GET /v2/elections?page=1&finalResults=true&titleSearchTerm=vocdoni&limit=20`
~~~js
{
 "elections": [...],
 "remainingPages": 9
}
~~~

And, as said above, fallback to the previous functionality (non-breaking) when no get params are received.

Following this approach, the current filtered POST method could also be updated by just adding  the new fields to the body:

`POST /elections/filter/page/:page`
~~~js
// Body
{
    "organizationId": "hexString",
    "electionId": "hexString",
    "withResults": false,
    "status": "READY",
    "limit": 20
}
// Response
{
 "elections": [...],
 "remainingPages": 9
}
~~~

Also, to avoid making this issue a burden, I'd try to split the tasks between the more interesting and less interesting ones, in terms of which ones we'll be using first. @selankon maybe you can help me here pin-point which ones we need first.

**Comment by elboletaire** [comment]
No more opinions? Does that mean yall ok with my last message? If so, I'd try to update the issue reflecting everything we've said.

Can we please prioritize the `total` or `remainingPages` or whatever-you-wanna-call-it property in all paginated responses?

**Comment by p4u** [comment]
After talking to @altergui we decided to go for the `GET /v2/elections?page=1&finalResults=true&titleSearchTerm=vocdoni&limit=20` while keeping backwards compatibility with the current endpoint. So no broken client.
